Allow user defined Call-ID to be sent on outgoing INVITE

austinkottke opened the pull request:
Modify the call setting structure to allow a custom Call-ID which does not depend on pjsip's dynamic pj_create_unique_string -- make_call will use the predefined Call-ID if specified. 

Instead it allows the app developer to have control. 

Example usage:

```
call_opt.custom_call_id = pj_str("0192983-28839-23287892");

pjsua_call_make_call(current_acc, &tmp, &call_opt, NULL, &msg_data_, &current_call);

```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pjsip/pjproject?pullRequest=4052) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pjsip/pjproject?pullRequest=4052) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pjsip/pjproject?pullRequest=4052) it.</sub>

austinkottke commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -1092,6 +1092,13 @@ typedef struct pjsua_call_setting
+     * Call-ID generated by dialog.
+     */
+
+    pj_str_t        user_call_id;
```
Changed to customCallId

austinkottke commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -1092,6 +1092,13 @@ typedef struct pjsua_call_setting
 
+    /***
+     * Custom Call-ID to be sent out with INVITE. Overrides default
+     * Call-ID generated by dialog.
```
Added your notes to documentation

austinkottke commented:
Commits have been added. Updated documentation to cpp api and c api to explicitly mention it is for outgoing calls only.

Changed the property to customCallId

sauwming commented on pjsip/src/pjsua-lib/pjsua_call.c:
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%s)", dlg->call_id->id  ));
```
`dlg->call_id->id` is of type `pj_str_t`. You need to use `%.*s` to print.

sauwming commented on pjsip/src/pjsua-lib/pjsua_call.c:
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%.*s)", (int)dlg->call_id->id.slen, dlg->call_id->id  ));
```
dlg->call_id->id.ptr

sauwming commented on pjsip/src/pjsua-lib/pjsua_call.c:
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
+     */
+
+    if( opt->custom_call_id.slen > 0 ){
+        dlg->call_id->id = opt->custom_call_id;
```
Need to duplicate the string using pj_strdup() as `opt` argument is const.

austinkottke replied:
Ok I see - thank you for pointing that out.

nanangizz commented on pjsip/src/pjsua-lib/pjsua_call.c:
```diff
@@ -644,6 +644,7 @@ PJ_DEF(void) pjsua_call_setting_default(pjsua_call_setting *opt)
     pj_bzero(opt, sizeof(*opt));
     opt->flag = PJSUA_CALL_INCLUDE_DISABLED_MEDIA;
     opt->aud_cnt = 1;
+    opt->custom_call_id = pj_str("");
```
minor optimization: unnecessary? as the `bzero()` above should set `pj_str_t.len` to zero already.

austinkottke replied:
Removed this. It appears unnecessary in my tests.

nanangizz commented on pjsip/include/pjsua-lib/pjsua.h:
```diff
@@ -1092,6 +1092,15 @@ typedef struct pjsua_call_setting
+     * Custom Call-ID to be sent out with outgoing INVITE. Overrides default
+     * Call-ID generated by dialog.
+     *
+     * This setting will only be used when creating a new outgoing call via pjsua_call_make_call().
```
- Better add info/warning that the library won't verify its validity (e.g: allowed chars, uniqueness).
- Column width should be < 80, for both, code & comment.
  Rather old school, but for consistency :)

austinkottke replied:
ok sounds good :)

nanangizz commented on pjsip/include/pjsua2/call.hpp:
```diff
@@ -333,6 +333,13 @@ struct CallSetting
+    /**
+     * User defined Call-ID to be sent out with outgoing INVITE
+     *
+     * This setting will only be used when creating a new outgoing call via Call::makeCall().
```
Similar as above in `pjsua_call_setting`:
- Info/warning about no validity check.
- Col width > 80?

nanangizz commented on pjsip/src/pjsua-lib/pjsua_call.c:
```diff
@@ -954,6 +955,16 @@ PJ_DEF(pj_status_t) pjsua_call_make_call(pjsua_acc_id acc_id,
+
+    if( opt->custom_call_id.slen > 0 ){
+        pj_strdup(dlg->pool, &dlg->call_id->id, &opt->custom_call_id);
+        PJ_LOG(4,(THIS_FILE, "Set user defined Call-ID (%.*s)", (int)dlg->call_id->id.slen, dlg->call_id->id.ptr  ));
```
col width > 80?
